### PR TITLE
Remove special cases from compose

### PR DESCRIPTION
**Pull request by jjant**
Removes extra logic to handle special cases from compose. Updated tests accordingly.

### CODE REVIEW

File: src/packages/recompose/compose.js [hunk 1]
```diff
@@ -1,11 +1,4 @@
...
-  if (funcs.length === 0) {
-    return arg => arg
-  }
-
-  if (funcs.length === 1) {
-    return funcs[0]
-  }
-
-  return funcs.reduce((a, b) => (...args) => a(b(...args)))
+  const identity = arg => arg
```
Comment (istarkov) [inline src/packages/recompose/compose.js]: can be moved out of function body

**Comment by TrySound** [comment]
Special cases are optiomisations. Let's keep it alive.

**Comment by TrySound** [comment]
And which edge cases?

File: src/packages/recompose/compose.js (hunk 1 above)
Comment (jjant) [inline src/packages/recompose/compose.js]: Right, thanks!

**Comment by jjant** [comment]
@TrySound Are those small optimisations really worth it? I think that this version is a bit cleaner, there's no need to add if statements to cover cases that `reduce` already handles.

**Comment by TrySound** [comment]
> Use to compose multiple higher-order components into a single higher-order component. This works exactly like the function of the same name in Redux, or lodash's flowRight().

**Comment by TrySound** [comment]
With your changs it becomes different.

**Comment by istarkov** [comment]
Just for one rare case

**Comment by jjant** [comment]
I don't see it, when does it become different?

**Comment by istarkov** [comment]
`expect(compose(fn)).toBe(fn)`

**Comment by jjant** [comment]
In that case you get the identity function composed with `fn` (i.e, `arg => fn(arg)`), which is really the same thing, you just lose reference equality, but I don't see how that's a problem.

**Comment by istarkov** [comment]
For me its also not a problem,  I think Ill accept this in near future.

**Comment by istarkov** [comment]
Just played with simple perf
like
```
var fn0 = (a) => a + 1;
var b = performance.now(); var zzz= compose_old(fn0, fn0, fn0, fn0, fn0, fn0); for(let i =0; i< 1000000; ++i) zzz(1); console.log(performance.now() - b);
// vs
var b = performance.now(); var zzz= compose_new(fn0, fn0, fn0, fn0, fn0, fn0); for(let i =0; i< 1000000; ++i) zzz(1); console.log(performance.now() - b);
```
and it gives that even for case with 6 fn0, the perf difference is near 25%, old version is faster.
Can be an issue as I use (and as I know not only I) compose from recompose for a lot of other cases.

**Pull request closed without merging**